libkatana_python_native/src/PropertyGraph.cpp

```diff
@@ -0,0 +1,742 @@
+
+template <typename T>
+py::object
+range(T start, T end, T step = 1) {
+  py::module builtins = py::module::import("builtins");
+  return builtins.attr("range")(start, end, step);
+}
+
+namespace {
+katana::GraphTopologyTypes::Edge
+out_edge_index(katana::PropertyGraph* pg, katana::GraphTopologyTypes::Node n) {
+  return *pg->OutEdges(n).begin();
+}
+katana::GraphTopologyTypes::Node
+out_edge_dst(katana::PropertyGraph* pg, katana::GraphTopologyTypes::Edge e) {
+  return *pg->OutEdgeDst(e);
+}
+}  // namespace
+
+void
+DefPropertyGraph(py::module& m) {
+  using namespace katana;
+  using namespace katana::python;
+
+  py::class_<PropertyGraph, std::shared_ptr<PropertyGraph>> cls(m, "Graph");
+
+  katana::DefCythonSupport(cls);
+  katana::DefConventions(cls);
+  katana::RegisterNumbaClass(cls);
+
+  cls.def(
+      py::init(
+          [](py::object path,
+             std::optional<std::vector<std::string>> node_properties,
+             std::optional<std::vector<std::string>> edge_properties,
+             TxnContext* txn_ctx) -> std::shared_ptr<PropertyGraph> {
+            auto path_str = py::str(path).cast<std::string>();
+            py::gil_scoped_release guard;
+            katana::RDGLoadOptions options = katana::RDGLoadOptions::Defaults();
+            options.node_properties = node_properties;
+            options.edge_properties = edge_properties;
+            TxnContextArgumentHandler txn_context_handler(txn_ctx);
+            KATANA_LOG_DEBUG("{}", reinterpret_cast<uintptr_t>(path.ptr()));
+            return PythonChecked(PropertyGraph::Make(
+                path_str, txn_context_handler.get(), options));
+          }),
+      py::arg("path"), py::arg("node_properties") = std::nullopt,
+      py::arg("edge_properties") = std::nullopt,
+      py::arg("txn_ctx") = TxnContextArgumentHandler::default_value,
+      R"""(
+      Load a property graph.
+
+      :param path: the path or URL from which to load the graph.
+      :type path: Union[str, Path]
+      :param node_properties: A list of node property names to load into memory. If this is None (default), then all
+          properties are loaded.
+      :param edge_properties: A list of edge property names to load into memory. If this is None (default), then all
+          properties are loaded.
+      )""");
+
+  // Below are the view API methods that we have in mind (not all of them may end up being exposed to Python). We differentiate between 3 classes of identifiers for nodes and edges:
+  // Local ID: This is an ID produced by a specific partitioning of a distributed graph.
+  // Topology handle: It is a handle used to relate to an entity in a topology data structure. It is a handle that various topology-related methods return.
+  // Property index: This is used to access that entity’s properties.
+  // We allow conversions between them in one direction: topology handle -> local ID -> property index.
+
+  katana::DefWithNumba<&PropertyGraph::NumNodes>(cls, "num_nodes");
```

Comment:
I always forget that we can actually pass references as template parameters...

Reply:
Actually you cannot pass references. What you can pass is pointers to functions (and maybe variables of specific kinds).

Reply:
Oh right, yes stupid mistake, you're passing in a function pointer. My bad.